packages/browser/core/src/core/auth/internal/StitchAuthImpl.ts

```diff
@@ -77,6 +110,98 @@ export default class StitchAuthImpl extends CoreStitchAuth<StitchUser>
     return super.loginWithCredentialInternal(credential);
   }
 
+  public loginWithRedirect(credential: StitchRedirectCredential): void {
+    const { redirectUrl, state } = prepareRedirect(
+      this,
+      credential,
+      RedirectTypes.Login
+    );
+
+    window.location.replace(
+      this.browserAuthRoutes.getAuthProviderRedirectRoute(
+        credential,
+        redirectUrl,
+        state,
+        this.deviceInfo
+      )
+    );
+  }
+
+  public hasRedirect(): boolean {
+    return !!this.storageRef.get(RedirectFragmentFields.State);
+  }
+
+  public handleRedirectResult(): Promise<StitchUser> {
+    const cleanup = () => {
+      this.storageRef.remove(RedirectFragmentFields.State);
+      this.storageRef.remove(RedirectKeys.RedirectProvider);
+    };
+
+    if (typeof window === "undefined") {
+      // This means we're running in some environment other
+      // than a browser - so handling a redirect makes no sense here.
+      cleanup();
+      return Promise.reject(
+        new StitchRedirectError(
+          "running in other than browser"  
+        )
+      );
+    }
+
+    if (!window.location || !window.location.hash) {
+      cleanup();
+      return Promise.reject(new StitchRedirectError("window location hash was undefined"));
+    }
+
+    const ourState = this.storageRef.get(RedirectFragmentFields.State);
+    const redirectProvider = this.storageRef.get(RedirectKeys.RedirectProvider);
+
+    cleanup();
+
+    const redirectFragment = window.location.hash.substring(1);
+    const redirectState = parseRedirectFragment(
+      redirectFragment,
+      ourState
+    );
+
+    if (redirectState.lastError || (redirectState.found && !redirectProvider)) {
+      // remove the fragment from the window history and reject
+      window.history.replaceState(null, "", pageRootUrl());
+      return Promise.reject(
+        new StitchRedirectError(
+          `StitchClient: error from redirect: ${
+            redirectState.lastError
+              ? redirectState.lastError
+              : "provider type not set"
+          }`
+        )
+      );
+    }
+
+    if (!redirectState.found) {
+      return Promise.reject(new StitchRedirectError(`no redirect state found`));
+    }
+
+    if (!redirectState.stateValid) {
+      window.history.replaceState(null, "", pageRootUrl());
+      return Promise.reject(new StitchRedirectError("StitchClient: state values did not match!"));
+    }
+
+    if (!redirectState.ua) {
+      return Promise.reject(new StitchRedirectError(
+        "StitchClient: no UA value was returned from redirect!"
+      ));
+    }
+
+    // If we get here, the state is valid - set auth appropriately.
+    return this.loginWithCredentialInternal(
+      new StitchAuthResponseCredential(redirectState.ua)
```

Comment:
I'm not sure this will completely work as written. You're missing the provider type and provider name, both of which need to be stored and retrieved (only one is stored). Please add a unit test that proves this.

Reply:
We should convert the ua here into the servers JSON form. An encode on ApiAuthInfo (toApi) would be sufficient and be the reverse to fromApi

Reply:
Sure that makes sense

Reply:
👍